pkg/apachepulsar/pulsar_source.go

```diff
@@ -0,0 +1,120 @@
+	pulsaradmin "github.com/streamnative/pulsar-admin-go"
+	"github.com/streamnative/pulsar-admin-go/pkg/utils"
+)
+
+type PulsarSource struct {
+	client      pulsar.Client
+	adminClient pulsaradmin.Client
+	consumer    pulsar.Consumer
+	toAckSet    map[string]pulsar.Message
+	lock        *sync.Mutex
+}
+
+func NewPulsarSource(client pulsar.Client, adminClient pulsaradmin.Client, consumer pulsar.Consumer) *PulsarSource {
+	return &PulsarSource{client: client, adminClient: adminClient, consumer: consumer, toAckSet: map[string]pulsar.Message{}, lock: new(sync.Mutex)}
+}
+
+func (ps *PulsarSource) Read(_ context.Context, readRequest sourcesdk.ReadRequest, messageCh chan<- sourcesdk.Message) {
+	ctx, cancel := context.WithTimeout(context.Background(), readRequest.TimeOut())
+	defer cancel()
+	// If we have un-acked data, we return without reading any new data.
+	if len(ps.toAckSet) > 0 {
+		return
+	}
+	for i := 0; i < int(readRequest.Count()); i++ {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			ps.lock.Lock()
+			msg, err := ps.consumer.Receive(ctx)
+			if err != nil {
+				log.Printf("error receiving message %s", err)
+			}
+			messageCh <- sourcesdk.NewMessage(
+				msg.Payload(),
+				sourcesdk.NewOffset([]byte(msg.ID().String()), 0),
+				msg.PublishTime(),
+			)
+			ps.toAckSet[msg.ID().String()] = msg
+			ps.lock.Unlock()
+		}
+	}
+}
+
+func (ps *PulsarSource) Pending(_ context.Context) int64 {
+	topic, _ := utils.GetTopicName(fmt.Sprintf("%s/%s/%s", os.Getenv("PULSAR_TENANT"), os.Getenv("PULSAR_NAMESPACE"), os.Getenv("PULSAR_TOPIC")))
+	stats, err := ps.adminClient.Topics().GetStats(*topic)
+	if err != nil {
+		log.Printf("error getting pending count %s", err)
+		return 0
+	}
+	pendingMessages := stats.Subscriptions[os.Getenv("PULSAR_SUBSCRIPTION")].UnAckedMessages
+	return pendingMessages
+}
+
+func (ps *PulsarSource) Ack(ctx context.Context, request sourcesdk.AckRequest) {
+	for _, offset := range request.Offsets() {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			func() {
+				ps.lock.Lock()
+				defer ps.lock.Unlock()
+				message := ps.toAckSet[string(offset.Value())]
+				err := ps.consumer.Ack(message)
+				if err != nil {
+					log.Printf("error in acknowledging message %s", err)
+					return
+				}
+				delete(ps.toAckSet, string(offset.Value()))
+			}()
+		}
+	}
+}
+
+func (ps *PulsarSource) Partitions(ctx context.Context) []int32 {
+	topic, err := utils.GetTopicName(fmt.Sprintf("%s/%s/%s", os.Getenv("PULSAR_TENANT"), os.Getenv("PULSAR_NAMESPACE"), os.Getenv("PULSAR_TOPIC")))
+	if err != nil {
+		log.Printf("error getting partitions from admin endpoint %s", err)
+		return sourcesdk.DefaultPartitions()
+	}
+	stats, err := ps.adminClient.Topics().GetMetadata(*topic)
+	if err != nil {
+		log.Printf("error getting partitions from admin endpoint %s", err)
+		return sourcesdk.DefaultPartitions()
+	}
+	var partitions []int32
+	for i := 0; i < stats.Partitions; i++ {
```

Comment:
This returns **global** partitions. The UDSource Partitions() API is supposed to return **local** partitions, aka. the partitions processed by the current pod. If a topic has 8 partitions, one pod could be processing only 2 of them.

In Kafka, each pod is associated with a session, and we use the session claim to retrieve partitions that the pod processes.  Does Apache Pulsar support a similar use case?

Reply:
In Apache Pulsar, the concept of partition assignment to specific consumers (or pods in a distributed setup) is somewhat different from Apache Kafka. Pulsar does not use the same concept of consumer sessions as Kafka does . I have gone through this official documentation -https://pulsar.apache.org/docs/3.1.x/concepts-messaging/#partitioned-topics ,and thats what I have figured out